api/envoy/api/v2/core/base.proto

```diff
@@ -95,6 +95,78 @@ message Metadata {
   map<string, google.protobuf.Struct> filter_metadata = 1;
 }
 
+// MetadataMatcher provides a general interface to check if a given value is matched in
+// :ref:`Metadata <envoy_api_msg_core.Metadata>`. It uses `filter` and `path` to retrieve the value
+// from the Metadata and then check if it's matched to one of the specified values.
+//
+// An example use of MetadataMatcher is specifying additional metadata in envoy.filters.http.rbac to
+// enforce access control based on dynamic metadata in a request.
+message MetadataMatcher {
+  // Specifies the value to match. Only primitive value is supported.
+  message Value {
+    // Specifies how to match a value. Only have effect on primitive value.
+    // For non-primitive value, it's always not matched.
+    oneof match_specifier {
+      option (validate.required) = true;
```

Comment:
instead of listing all combination of type and its matcher here, would it be clearer to make a Matcher type (with oneof) for each type and have a oneof here?

i.e.:

```
message NullMatch {
}

message NumberMatch {
  oneof match {
    Int64Range range_match = 1;
    int64 exact_match = 2;
  }
}

message ValueMatch {
  oneof match {
    NullMatch null = 1;
    NumberMatch number = 2;
    StringMatch string = 3;
    ...
  }
}
```

Reply:
I think that makes sense for string, number or other non primitive type, for bool I just kept it there as it's much simpler to use and also easy to compare in the code.